controllers/services/monitoring/monitoring_controller_actions.go

```diff
@@ -5,44 +5,109 @@ import (
+	// Map component names to their rule prefixes
+	dscList := &dscv1.DataScienceClusterList{}
+	if err := rr.Client.List(ctx, dscList); err != nil {
+		return fmt.Errorf("failed to list DSC: %w", err)
+	}
+	if len(dscList.Items) == 0 {
+		return nil
+	}
+	dsc := &dscList.Items[0]
+	componentRules := map[string]string{
+		componentApi.DashboardComponentName:            "rhods-dashboard",
+		componentApi.WorkbenchesComponentName:          "workbenches",
+		componentApi.KueueComponentName:                "kueue",
+		componentApi.CodeFlareComponentName:            "codeflare",
+		componentApi.DataSciencePipelinesComponentName: "data-science-pipelines-operator",
+		componentApi.ModelMeshServingComponentName:     "model-mesh",
+		componentApi.RayComponentName:                  "ray",
+		componentApi.TrustyAIComponentName:             "trustyai",
+		componentApi.KserveComponentName:               "kserve",
+		componentApi.TrainingOperatorComponentName:     "trainingoperator",
+		componentApi.ModelRegistryComponentName:        "model-registry-operator",
+		componentApi.ModelControllerComponentName:      "odh-model-controller",
 	}
 
-	// url
-	rl := routev1.RouteList{}
-	err := rr.Client.List(
-		ctx,
-		&rl,
-		client.InNamespace(rr.DSCI.Spec.Monitoring.Namespace),
-		client.MatchingLabels(map[string]string{
-			labels.PlatformPartOf: serviceApi.MonitoringServiceName,
-		}),
-	)
+	err := cr.ForEach(func(ch cr.ComponentHandler) error {
+		var enabled bool
+		ci := ch.NewCRObject(dsc)
+		// read the component instance to get tha actual status
+		err := rr.Client.Get(ctx, client.ObjectKeyFromObject(ci), ci)
+		switch {
+		case k8serr.IsNotFound(err):
+			enabled = false
+		case err != nil:
+			enabled = false
+			return fmt.Errorf("error getting component state component: %v enabled: %v %w ", ch.GetName(), enabled, err)
+		default:
+			enabled = meta.IsStatusConditionTrue(ci.GetStatus().Conditions, status.ConditionTypeReady)
+		}
 
+		// Check for shared components
+		if ch.GetName() == componentApi.KserveComponentName || ch.GetName() == componentApi.ModelMeshServingComponentName {
+			if err := UpdatePrometheusConfig(ctx, enabled, componentRules[componentApi.ModelControllerComponentName]); err != nil {
+				return err
+			}
+		}
+
+		if err := UpdatePrometheusConfig(ctx, enabled, componentRules[ch.GetName()]); err != nil {
+			return err
+		}
+		return nil
+	})
 	if err != nil {
-		return fmt.Errorf("failed to list routes: %w", err)
+		return err
 	}
+	return nil
+}
 
-	d.Status.URL = ""
-	if len(rl.Items) == 1 {
-		d.Status.URL = resources.IngressHost(rl.Items[0])
+func updateStatus(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	m, ok := rr.Instance.(*serviceApi.Monitoring)
+	if !ok {
+		return errors.New("instance is not of type *services.Monitoring")
 	}
 
+	m.Status.Phase = "Ready"
+	m.Status.ObservedGeneration = m.GetObjectMeta().GetGeneration()
```

Comment:
ideally we should also have a `Ready` condition

Reply:
add an NoReady case, only set to Ready when Monitoring Prom deployment is ready

Reply:
I meant a ready **Condition**, phase is not that much in use

Reply:
sorry, did not see this, let me get a new one next week for this comment